### PR TITLE
enable fraud service to full & light nodes and extend abci.FraudProof with go-fraud Proof interface

### LINKED ISSUE #892: Add `StateFraudProof` proof type

### PR DESCRIPTION
**Pull request by gupadhyaya**
As part of go-fraud library implementation, this PR adds:
* ProofService to full and light nodes
* Extends abci.FraudProof interface with go-fraud Proof interface

Fixes #892 #893 #894

### CODE REVIEW

File: node/proof_service_factory.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+		client:        c,
+		getter:        getter,
+		ds:            ds,
+		syncerEnabled: syncerEnabled,
+		proofType:     proofType,
+	}
+}
+
+// OnStart is a part of Service interface.
+func (factory *ProofServiceFactory) Start() *fraudserv.ProofService {
```
Comment (tzdybal) [inline node/proof_service_factory.go:30]: [optional, nitpicking] calling `factory.Start()` looks like starting factory, not instantiating object by the factory.
`Create` seems more natural.

**Review by tzdybal — CHANGES_REQUESTED** [review]
This new code needs tests.

File: node/proof_service_factory.go (hunk 1 above)
Comment (Manav-Aggarwal) [reply node/proof_service_factory.go:30]: Agreed since `Start` is not part of the `Service` interface. Maybe even `CreateProofService` if we want to be more explicit.

File: node/header_exchange.go [hunk 2]
```diff
@@ -218,7 +218,7 @@ func newP2PExchange(
 ) (*goheaderp2p.Exchange[*types.SignedHeader], error) {
 	opts = append(opts,
 		goheaderp2p.WithNetworkID[goheaderp2p.ClientParameters](network),
-		goheaderp2p.WithChainID[goheaderp2p.ClientParameters](chainID),
+		goheaderp2p.WithChainID(chainID),
```
Comment (Manav-Aggarwal) [inline node/header_exchange.go:221]: How is this change related to this PR?

**Review by Manav-Aggarwal — CHANGES_REQUESTED** [review]
Thanks for the PR! Left some comments.

File: node/header_exchange.go (hunk 2 above)
Comment (gupadhyaya) [reply node/header_exchange.go:221]: unrelated, but there was a warning, hence went ahead and fixed it.

**Comment by gupadhyaya** [comment]
> This new code needs tests.
this PR only adds fraud-exchange service to full and light nodes and for that the graceful start and stop is already tested in other unit tests. the fraud-exchange functionality will be added in the next PR along with new tests to cover the exchange part. will that suffice?

File: node/proof_service_factory.go (hunk 1 above)
Comment (gupadhyaya) [reply node/proof_service_factory.go:30]: agreed.

File: node/proof_service_factory.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	return ProofServiceFactory{
+		client:        c,
+		getter:        getter,
+		ds:            ds,
+		syncerEnabled: syncerEnabled,
+		proofType:     proofType,
+	}
+}
+
+// OnStart is a part of Service interface.
```
[the remark this answers is not in the document]
Comment (gupadhyaya) [reply node/proof_service_factory.go:29]: not really. removed.

File: types/fraud_proof.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	return uint64(fp.BlockHeight)
+}
+
+func (fp *FraudProof) Validate(header.Header) error {
+	// TODO (ganesh): fill this later
+	return nil
+}
+
+func (fp *FraudProof) MarshalBinary() (data []byte, err error) {
+	return json.Marshal(fp)
```
[the remark this answers is not in the document]
Comment (gupadhyaya) [reply types/fraud_proof.go]: fixed by calling the Marshal/Unmarshal from abci.Fraud

File: node/full.go [hunk 5]
```diff
@@ -161,27 +166,38 @@ func newFullNode(
 		return nil, fmt.Errorf("HeaderExchangeService initialization error: %w", err)
 	}
 
+	fraudProofFactory := NewProofServiceFactory(
+		client,
+		func(ctx context.Context, u uint64) (header.Header, error) {
+			return headerExchangeService.headerStore.GetByHeight(ctx, u)
+		},
+		mainKV,
+		true,
```
[the remark this answers is not in the document]
Comment (gupadhyaya) [reply node/full.go:175]: the ProofServiceFactory requires the `syncerEnabled` parameter and i think we can instantiate a ProofService with syncer disabled in the tests. For now, we can keep it and based on the 2nd PR that utilizes the ProofService, i can consolidate it if not needed.

**Review by tzdybal — APPROVED** [review]

**Review by tzdybal — APPROVED** [review]
